### DISCUSSION TITLE
Recursive type is considered generic

### DISCUSSION THREAD

**Discussion by notEvil** [microsoft/pyright#6614] [Q&A] [answered]
Hi,

Pyright 1.1.338 resolves `Any` in

```python
import collections.abc as c_abc

List = c_abc.Sequence[typing.TypeVar("T", bound="Any")]
Set = c_abc.Set[typing.TypeVar("T", bound="Any")]
Map = c_abc.Mapping[
    typing.TypeVar("KT", bound="Any"), typing.TypeVar("VT", bound="Any")
]
Any = bool | List["Any"] | Set["Any"] | Map["Any", "Any"]

typing.TypeVar("_", bound=Any)
```

as being generic ("TypeVar bound type cannot be generic") in `List = ...` and the last line. Is this intended?

Some modifications where it is not considered generic:

<details>
<summary>Single type of container</summary>

```python
import collections.abc as c_abc

List = c_abc.Sequence[typing.TypeVar("T", bound="Any")]
Any = bool | List["Any"]

typing.TypeVar("_", bound=Any)
```

```python
import collections.abc as c_abc

Set = c_abc.Set[typing.TypeVar("T", bound="Any")]
Any = bool | Set["Any"]

typing.TypeVar("_", bound=Any)
```

```python
import collections.abc as c_abc

Map = c_abc.Mapping[
    typing.TypeVar("KT", bound="Any"), typing.TypeVar("VT", bound="Any")
]
Any = bool | Map["Any", "Any"]

typing.TypeVar("_", bound=Any)
```

</details>

<details>
<summary>Single generic container</summary>

```python
import collections.abc as c_abc

List = c_abc.Sequence[typing.TypeVar("T", bound="Any")]
Set = c_abc.Set
Map = c_abc.Mapping
Any = bool | List["Any"] | Set["Any"] | Map["Any", "Any"]

typing.TypeVar("_", bound=Any)
```

```python
import collections.abc as c_abc

List = c_abc.Sequence
Set = c_abc.Set[typing.TypeVar("T", bound="Any")]
Map = c_abc.Mapping
Any = bool | List["Any"] | Set["Any"] | Map["Any", "Any"]

typing.TypeVar("_", bound=Any)
```

```python
import collections.abc as c_abc

List = c_abc.Sequence
Set = c_abc.Set
Map = c_abc.Mapping[
    typing.TypeVar("KT", bound="Any"), typing.TypeVar("VT", bound="Any")
]
Any = bool | List["Any"] | Set["Any"] | Map["Any", "Any"]

typing.TypeVar("_", bound=Any)
```

</details>

In the actual application, `List`, `Set` and `Map` are generic unions with 2 and 4 typevars, so I can't really get rid of the typevars. And except for this "issue", Pyright seems to handle them pretty well.

**Comment by erictraut** [comment, marked as answer]
Pyright is working as designed here, so I don't consider this a bug.

Traditional (implicit) type aliases use the same syntax as a variable assignment. Type checkers need to use various heuristics to determine whether it should treat such a statement as a type alias. One of the heuristics that pyright uses is that the expression on the righthand side must be a valid type expression. Function calls are not allowed in type expressions. You're calling `typing.TypeVar` in these expressions, so pyright is not treating them as type aliases.

TypeVars should be defined in the global (module) scope, not allocated inline within a type annotation. If you change your code accordingly, it will work fine.

```python
import typing
import collections.abc as c_abc

T = typing.TypeVar("T", bound="Any")
KT = typing.TypeVar("KT", bound="Any")
VT = typing.TypeVar("VT", bound="Any")

List = c_abc.Sequence[T]
Set = c_abc.Set[T]
Map = c_abc.Mapping[KT, VT]
Any = bool | List["Any"] | Set["Any"] | Map["Any", "Any"]

typing.TypeVar("_", bound=Any)
```

I recommend against using old-style (implicit) type alias definitions. The recommended approach is to use PEP 613-style `TypeAlias` declarations if you're on Python 3.11 or older or to use the new PEP 695-style `type` statement if you're on Python 3.12.

Here's how it looks with the PEP 695 syntax.

```python
type List[T: "Any"] = c_abc.Sequence[T]
type Set[T: "Any"] = c_abc.Set[T]
type Map[KT: "Any", VT: "Any"] = c_abc.Mapping[KT, VT]
type Any = bool | List["Any"] | Set["Any"] | Map["Any", "Any"]
```

**Reply by notEvil** [reply]
Thanks a lot! It never crossed my mind that just calling `typing.TypeVar` wouldn't be the same as assign(name, call) + load(name).
